### PR TITLE
Socrata export ETL

### PR DESCRIPTION
**Pull request by johnclary**
## Associated issues

- https://github.com/cityofaustin/atd-data-tech/issues/18149

This is the new Socrata export! The diff is not as bad as it looks—I had to renumber a bunch of migrations.

These follow-up tasks will be addressed in separate issues/PRs:

- airflow
- docker image CI
- prepare/coordinate updates to the production datasets to make them match the changes made here

## Testing

You can run this using a local db, but we're going to use the test instance cluster since it is already loaded up with 10 years of data.

### Setup

1. In our password store, locate the item called `Env file for the Vision Zero new data model Socrata export ETL` and save it in the root of this ETL directory as `.env`.

2. Build and run the Socrata export. This may take up to 20 minutes.

```shell
# from ./atd-etl/socrata_export
$ docker compose run socrata_export
$ ./socrata_export.py --crashes --people
```

3. Nice! Start your local VZV and inspect the summary page and map

```
# from ./atd-vzv
$ nvm use
$ npm run start
```

4. Test out the helper script, which logs the differences between the staging datasets and production datasets, which will need to be brought into sync with the staging datasets when we launch.

```shell
# from ./atd-etl/socrata_export
$ docker compose run socrata_export
$ ./_compare_column_metadata.py
```

### CODE REVIEW

File: atd-etl/socrata_export/_compare_column_metadata.py
Comment (johnclary) [inline atd-etl/socrata_export/_compare_column_metadata.py:1]: This script identifies the differences between the staging and prod datasets and prints it in a nice way, like this:

**** crashes dataset changes ****

- change `crash_fatal_fl` from type `text` to `checkbox` in the prod dataset
- change `onsys_fl` from type `text` to `checkbox` in the prod dataset
- change `private_dr_fl` from type `text` to `checkbox` in the prod dataset
- change `road_constr_zone_fl` from type `text` to `checkbox` in the prod dataset
- create column `address_primary` type `text` in the prod dataset
- create column `address_secondary` type `text` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `law_enf_fatality_count` type `number` in the prod dataset
- create column `vz_fatality_count` type `number` in the prod dataset
- delete column `apd_confirmed_death_count` from the prod dataset
- delete column `apd_confirmed_fatality` from the prod dataset
- delete column `bicycle_fl` from the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `micromobility_fl` from the prod dataset
- delete column `motor_vehicle_fl` from the prod dataset
- delete column `motorcycle_fl` from the prod dataset
- delete column `other_fl` from the prod dataset
- delete column `pedestrian_fl` from the prod dataset
- delete column `rpt_latitude` from the prod dataset
- delete column `rpt_longitude` from the prod dataset
- delete column `street_name_2` from the prod dataset
- delete column `street_name` from the prod dataset
- delete column `street_nbr_2` from the prod dataset
- delete column `street_nbr` from the prod dataset

**** people dataset changes ****

- change `mode_id` from type `text` to `number` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `is_primary_person` type `checkbox` in the prod dataset
- create column `prsn_ethnicity_label` type `text` in the prod dataset
- create column `prsn_sex_id` type `number` in the prod dataset
- create column `prsn_sex_label` type `text` in the prod dataset
- create column `unit_id` type `number` in the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `prsn_gndr_id` from the prod dataset

File: atd-toolbox/data_model/python/create_migrations.sh
Comment (johnclary) [inline atd-toolbox/data_model/python/create_migrations.sh:1]: i finally created this helper to run all the python sql-gen scripts 👍

File: atd-toolbox/data_model/python/40_make_lookup_seeds.py [hunk 1]
```diff
@@ -15,8 +16,8 @@ def escape_single_quotes(input_string):
 veh_direction_of_force_lkp_values = """insert into lookups.veh_direction_of_force_lkp (id, label)
     select veh_direction_of_force_id, veh_direction_of_force_desc from atd_txdot__veh_direction_of_force_lkp;"""
 
-mode_category_lkp_values = """insert into lookups.mode_category_lkp (id, label)
-    select  id, atd_mode_category_mode_name from atd__mode_category_lkp;"""
+mode_category_lkp_values = """insert into lookups.mode_category_lkp (id, label, source)
+    select  id, atd_mode_category_mode_name, 'vz' from atd__mode_category_lkp where id != 10;"""
```
Comment (johnclary) [inline atd-toolbox/data_model/python/40_make_lookup_seeds.py:20]: making it super clear that this lookup table is completely custom by adding a `source` column and setting it to `vz`. (we use the `source` column in other lookup tables which mix CRIS + VZ values—we have a documentation todo about this 👍 )

File: atd-toolbox/data_model/python/column_metadata.json
Comment (johnclary) [inline atd-toolbox/data_model/python/column_metadata.json:1]: renamed `atd_mode_category ` to `vz_mode_category_id`

File: atd-toolbox/data_model/python/column_metadata.json [hunk 2]
```diff
@@ -2262,19 +2262,19 @@
...
+    "column_name": "vz_mode_category_id",
     "data_type": "integer",
     "data_type_constraint": "",
     "foreign_column_cascade": "on update cascade on delete cascade",
     "foreign_column_name": "id",
     "foreign_table_name": "mode_category_lkp",
     "foreign_table_schema": "lookups",
     "is_cris_column": false,
-    "is_edits_column": true,
+    "is_edits_column": false,
```
Comment (johnclary) [inline atd-toolbox/data_model/python/column_metadata.json:2273]: it's a generated column, so we don't want it in the `_edits` schema

File: atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql
Comment (johnclary) [inline atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:1]: summary of changes to this file:
- fix up some joining on `crashes.id` versus `crashes.crash_id`
- add the mode-based injury metric columns, which are rather tedious but we use them in the VZV
- use `LATERAL JOIN`s in the `crashes_list_view`, which turns out to be the silver bullet for keeping these queries as efficient as possible

File: atd-vze/.env.test
Comment (johnclary) [inline atd-vze/.env.test:1]: new env configuration for the test instance 🧑‍🔬

File: atd-etl/socrata_export/socrata_export.py [hunk 3]
```diff
@@ -1,121 +1,95 @@
...
-            raise RuntimeError("There was a problem gathering data.")
+        logger.info(f"Processing {dataset_name}")
 
-        # Format records
-        records = config["formatter"](data, config["formatter_config"])
+        min_id = 0
+        records_processed = 0
+        while True:
+            logger.info(f"Fetching up to {RECORD_BATCH_SIZE} records")
+            variables = {"limit": RECORD_BATCH_SIZE, "minId": min_id}
```
Comment (johnclary) [inline atd-etl/socrata_export/socrata_export.py:48]: i initially implemented this using the `offset` flag, but it was extremely slow. querying by `id` with a record limit is much faster. 

interestingly, i also had to push the `order by id` clause down into the views. this was dramatically faster than querying the view with an `order` directive. it's a limitation of the pg query planner that i don't understand.

File: atd-etl/socrata_export/socrata_export.py
Comment (johnclary) [inline atd-etl/socrata_export/socrata_export.py:1]: as we're doing today, this script truncates and reloads each dataset in full every time it runs. that's ~150,000 crash records nightly 😵 

i think we could switch the crash level dataset over to incremental loading and it would work reliably. the person-level data is a bit more challenging because we only publish records for serious injuries and fatalities, so we would need a mechanism to remove records from person dataset. we could also switch over to publishing all person records. regardless of how we implement incremental loading, i think we would need some intermittent process that does a full truncate/replace to make sure we're 100% in sync.

TLDR i don't think we should fuss with incremental loading for the time being.

File: atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql
Comment (johnclary) [inline atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:1]: summary of changes to this file:
- fix up some joining on `crashes.id` versus `crashes.crash_id`
- order views by `id`
- add the mode-based injury metric columns, which are rather tedious but we use them in the VZV
- use `LATERAL JOIN`s in the `crashes_list_view`, which turns out to be the silver bullet for keeping these queries fast

File: atd-vze/.env.test
Comment (johnclary) [inline atd-vze/.env.test:1]: new env configuration for the test instance 🧑‍🔬

File: atd-etl/socrata_export/_compare_column_metadata.py
Comment (johnclary) [inline atd-etl/socrata_export/_compare_column_metadata.py:1]: the output of this script looks like this:

```**** crashes dataset changes ****

- change `crash_fatal_fl` from type `text` to `checkbox` in the prod dataset
- change `onsys_fl` from type `text` to `checkbox` in the prod dataset
- change `private_dr_fl` from type `text` to `checkbox` in the prod dataset
- change `road_constr_zone_fl` from type `text` to `checkbox` in the prod dataset
- create column `address_primary` type `text` in the prod dataset
- create column `address_secondary` type `text` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `id` type `number` in the prod dataset
- create column `law_enf_fatality_count` type `number` in the prod dataset
- delete column `apd_confirmed_death_count` from the prod dataset
- delete column `apd_confirmed_fatality` from the prod dataset
- delete column `atd_mode_category_metadata` from the prod dataset
- delete column `bicycle_fl` from the prod dataset
- delete column `contrib_factr_p1_id` from the prod dataset
- delete column `contrib_factr_p2_id` from the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `micromobility_fl` from the prod dataset
- delete column `motor_vehicle_fl` from the prod dataset
- delete column `motorcycle_fl` from the prod dataset
- delete column `other_fl` from the prod dataset
- delete column `pedestrian_fl` from the prod dataset
- delete column `rpt_latitude` from the prod dataset
- delete column `rpt_longitude` from the prod dataset
- delete column `street_name_2` from the prod dataset
- delete column `street_name` from the prod dataset
- delete column `street_nbr_2` from the prod dataset
- delete column `street_nbr` from the prod dataset

**** people dataset changes ****

- change `mode_id` from type `text` to `number` in the prod dataset
- create column `crash_timestamp_ct` type `calendar_date` in the prod dataset
- create column `crash_timestamp` type `calendar_date` in the prod dataset
- create column `id` type `number` in the prod dataset
- create column `is_primary_person` type `checkbox` in the prod dataset
- create column `prsn_ethnicity_label` type `text` in the prod dataset
- create column `prsn_sex_id` type `number` in the prod dataset
- create column `prsn_sex_label` type `text` in the prod dataset
- create column `unit_id` type `number` in the prod dataset
- delete column `crash_date` from the prod dataset
- delete column `crash_time` from the prod dataset
- delete column `prsn_gndr_id` from the prod dataset
```

File: atd-etl/socrata_export/_compare_column_metadata.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+]
+
+
+def compare_col_arrays(list_a, list_b, a_list_name, b_list_name, primary_list="a"):
+    todos = []
+    for col in list_a:
+        key = col["fieldName"]
+        matching_col = None
+        try:
+            matching_col = next(
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/_compare_column_metadata.py:18]: this is a cool way to do this, lots of TILs in this python script for me :cowboy_hat_face:

File: atd-etl/socrata_export/socrata_export.py [hunk 5]
```diff
@@ -1,121 +1,95 @@
...
-                "veh_body_styl_desc": "unit_desc",
-                "veh_unit_desc_desc": "unit_mode",
-                "latitude_primary": "latitude",
-                "longitude_primary": "longitude",
-                "atd_fatality_count": "death_cnt",
-            },
-        },
-        "dataset_uid": ATD_ETL_CONFIG["SOCRATA_DATASET_CRASHES"],
+        "name": "crashes",
+        "typename": "socrata_export_crashes_view",
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/socrata_export.py:21]: lovinggg the new socrata views

File: atd-etl/socrata_export/socrata_export.py [hunk 6]
```diff
@@ -1,121 +1,95 @@
...
-        if records and not is_truncate_complete:
-            print(f'Truncating {config["table"]} table...')
-            client.replace(config["dataset_uid"], [])
-            is_truncate_complete = True
+            if min_id == 0:
+                if not records:
+                    raise Exception(
+                        "No records returned from Hasura. This should never happen"
+                    )
+                # we have some records, so it's safe to truncate the Socrata dataset
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/socrata_export.py:58]: by truncate you mean dropping all the data in the table and replacing it?

File: atd-etl/socrata_export/socrata_export.py [hunk 7]
```diff
@@ -1,121 +1,95 @@
...
+                socrata_client.replace(dataset["dataset_id"], [])
+            elif not records:
+                logger.info(f"{records_processed} {dataset_name} records processed")
+                break
 
-        # Upsert records to Socrata
-        client.upsert(config["dataset_uid"], records)
-        total_records += len(records)
+            logger.info(
+                f"Upserting {len(records)} records IDs: {records[0]['id']} - {records[-1]['id']} (total records: {records_processed})"
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/socrata_export.py:66]: are we upserting tho or just completely replacing?

File: atd-etl/socrata_export/socrata_export.py [hunk 8]
```diff
@@ -1,121 +1,95 @@
...
 
-        if "data" not in data:
-            print("query: ", query)
-            print("data: ", data)
-            raise RuntimeError("There was a problem gathering data.")
+        logger.info(f"Processing {dataset_name}")
 
-        # Format records
-        records = config["formatter"](data, config["formatter_config"])
+        min_id = 0
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/socrata_export.py:44]: this code took me some time to understand, could you add some comments describing whats going on -- my understanding is we are looping thru querying the hasura database 1000 records at a time starting with a min record ID which gets incremented by +1 of the last record id at the end of the 1000 record batch replace. this was making me :woozy_face: trying to understand but maybe thats just me

File: atd-etl/socrata_export/socrata_export.py [hunk 9]
```diff
@@ -1,121 +1,95 @@
...
-            client.replace(config["dataset_uid"], [])
-            is_truncate_complete = True
+            if min_id == 0:
+                if not records:
+                    raise Exception(
+                        "No records returned from Hasura. This should never happen"
+                    )
+                # we have some records, so it's safe to truncate the Socrata dataset
+                logger.info(f"Truncating dataset {dataset['dataset_id']}")
+                socrata_client.replace(dataset["dataset_id"], [])
```
Comment (roseeichelmann) [inline atd-etl/socrata_export/socrata_export.py:60]: im confused, why are we replacing the dataset when min_id = 0 and then afterwards we are upserting?

File: atd-etl/socrata_export/socrata_export.py (hunk 7 above)
Comment (roseeichelmann) [reply atd-etl/socrata_export/socrata_export.py:66]: jk i see we are upserting later i just got confused by how we replace at first?

File: atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql [hunk 10]
```diff
@@ -140,10 +261,46 @@ create or replace view crash_injury_metrics_view as
...
+        ), 0) as other_fatality_count,
+        coalesce(sum(
+            person_injury_metrics_view.other_sus_serious_injry
+        ), 0) as other_sus_serious_injry_count,
         case
             when (sum(person_injury_metrics_view.fatal_injury) > 0) then 4
-            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 1
-            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.nonincap_injry) > 0) then 2
+            when (sum(person_injury_metrics_view.sus_serious_injry) > 0) then 1
```
Comment (roseeichelmann) [inline atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:303]: dont we want the `crash_injry_sev_id` to be representative of the most severe injury involved in the crash? so fatality -> serious injury -> nonincap injury etc? or am i misunderstanding this field

File: atd-toolbox/data_model/python/sql_templates/people_list_view.sql [hunk 11]
```diff
@@ -21,5 +21,5 @@ create or replace view people_list_view as (
         on lookups.injry_sev_lkp.id = people.prsn_injry_sev_id
     left join
         lookups.mode_category_lkp
-        on units.unit_desc_id = lookups.mode_category_lkp.id
+        on units.vz_mode_category_id = lookups.mode_category_lkp.id
```
Comment (roseeichelmann) [inline atd-toolbox/data_model/python/sql_templates/people_list_view.sql:24]: great change :+1:

File: atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+left join lateral (
+    select *
+    from
+        unit_aggregates
+    where
+        crashes.id = unit_aggregates.id
+    limit 1
+) as unit_aggregates on true
+where
+    crashes.in_austin_full_purpose = true and crashes.private_dr_fl = false
```
Comment (roseeichelmann) [inline atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql:88]: should we be using this `private_dr_fl` like in the fatalities view as well?

**Review by roseeichelmann — APPROVED** [review]
this all looks soooo great john. tested with all your steps, it all seems to be working as expected. the helper script to compare the columns is a great touch. i just have a few questions and the socrata export script was a little hard for me to follow but i think i figured it out pretty well after taking time to comb into the code. exciting stuff!!! :rocket:

File: atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql (hunk 10 above)
Comment (chiaberry) [inline atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:303]: Im trying to follow whats happening, why did these swap?

Comment (chiaberry) [reply atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:303]: (I didnt see Rose's comment before I left mine, I need to refresh my tabs when im taking super long to review code)

File: atd-etl/socrata_export/_compare_column_metadata.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+            continue
+        # see if the column type is the same
+        a_col_type = col["dataTypeName"]
+        b_col_type = matching_col["dataTypeName"]
+        if a_col_type != b_col_type:
+            if primary_list == "a":
+                todo = f"- change `{key}` from type `{b_col_type}` to `{a_col_type}` in the {b_list_name} dataset"
+                todos.append(todo)
+            else:
+                # we don't need to print these type mismatches
```
Comment (chiaberry) [inline atd-etl/socrata_export/_compare_column_metadata.py:39]: why doesnt it matter if the datatypes are different when the primary_list is b?

File: atd-etl/socrata_export/_compare_column_metadata.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def main():
+    socrata_client = get_socrata_client()
+    for dataset in datasets:
+        print(f"\n**** {dataset['name']} dataset changes ****\n")
+        cols_staging = socrata_client.get_metadata(dataset["staging"])["columns"]
+        cols_prod = socrata_client.get_metadata(dataset["prod"])["columns"]
+        todos = compare_col_arrays(cols_staging, cols_prod, "staging", "prod", primary_list="a")
+        todos += compare_col_arrays(cols_prod, cols_staging, "staging", "prod", primary_list="b")
```
Comment (chiaberry) [inline atd-etl/socrata_export/_compare_column_metadata.py:53]: I'm confused here

so the list_a and list_b lists are swapped, cols_prod is now list_a (when the line above here, cols_prod was list_b)

*but* a_list_name and b_list_name aren't swapped, so is it more like primary_list_name and secondary_list_name?

File: atd-etl/socrata_export/socrata_export.py (hunk 5 above)
Comment (chiaberry) [reply atd-etl/socrata_export/socrata_export.py:21]: +1

File: atd-etl/socrata_export/socrata_export.py (hunk 7 above)
Comment (chiaberry) [reply atd-etl/socrata_export/socrata_export.py:66]: When min_id is 0 it replaces the whole dataset with an empty `[]` , and then does an upsert on every run until we run out of records, which is why the break happens on line 63. 

Could the logger say "(records already processed: {records_processed}" instead of "total records"? since its not the total number of records. 
`Upserting 1000 records IDS: 1-1000 (total records: 0)` is what I anticipate will log on the first run, right?

**Comment by chiaberry** [comment]
I didn't run it, but read through the code and left a question or two. I got tripped up with the compare_column_metadata script a bit, but nothing major I don't think.

File: atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    and crashes.crash_timestamp < now() - interval '14 days' order by id asc;
```
Comment (mddilley) [inline atd-toolbox/data_model/python/sql_templates/socrata_export_views.sql:89]: love having this interval tucked into the database view. I know we touched on it long ago but it is cool to see this in action. Hooray for filtering in one place 🎉

**Review by mddilley — CHANGES_REQUESTED** [review]
I tested this with the VZ test cluster [details here](https://austininnovation.slack.com/archives/CM9MK950S/p1720561270538219?thread_ts=1720559497.280309&cid=CM9MK950S), and I was able to fully replace both crash and people test datasets. I saw no obvious regressions in my local VZV either. It is so nice to see these refactored into database views. I know this will be more maintainable, and the sql really spells out the business logic well.

I found the tiniest typo that held up me doing a local run successfully, and I left some questions but I think this is 🚢 after the fix. 🙌

File: atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql (hunk 10 above)
Comment (johnclary) [reply atd-toolbox/data_model/python/sql_templates/crashes_list_view.sql:303]: ah yes—they swapped because injury severity ID `1`  = `sus_serious_injry`. so that was a bug. but as @roseeichelmann pointed out i introduced a new bug in my `case` statement 🙏

File: atd-etl/socrata_export/_compare_column_metadata.py (hunk 13 above)
Comment (johnclary) [reply atd-etl/socrata_export/_compare_column_metadata.py:39]: we only care about making the prod dataset look like the staging dataset, not the other way around. so there is no need to print anything. i am sorry this code is not easier to grok 😞

File: atd-etl/socrata_export/_compare_column_metadata.py (hunk 14 above)
Comment (johnclary) [reply atd-etl/socrata_export/_compare_column_metadata.py:53]: 🍝

File: atd-etl/socrata_export/socrata_export.py (hunk 8 above)
Comment (johnclary) [reply atd-etl/socrata_export/socrata_export.py:44]: will do! and yes, you got it :)

File: atd-etl/socrata_export/_compare_column_metadata.py (hunk 14 above)
Comment (johnclary) [reply atd-etl/socrata_export/_compare_column_metadata.py:53]: i'll touch up this code to try to make var names clearer

File: atd-etl/socrata_export/socrata_export.py (hunk 7 above)
Comment (johnclary) [reply atd-etl/socrata_export/socrata_export.py:66]: Thanks for suggesting the logging improvements. And I will add some comments to the code that make the logic for querying records and replacing the dataset more clear 🙏

File: atd-etl/socrata_export/utils/graphql.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import os
+
+import requests
+
+ENDPOINT = os.getenv("HASURA_GRAPHQL_ENDPOINT")
+ADMIN_SECRET = os.getenv("HASURA_GRAPQL_ADMIN_SECRET")
```
[the remark this answers is not in the document]
Comment (johnclary) [reply atd-etl/socrata_export/utils/graphql.py]: nice catch—thank you! ✅

**Comment by johnclary** [comment]
👋 thanks for the feedback. i reworked the logging and added docstrings in `socrata_export.py` and i refactored `_compare_column_metadata.py` to be less confusing 👍

**Review by mddilley — APPROVED** [review]
I just ran and replaced the crashes and people datasets again, and all worked perfectly!  I also diffed the output of the compare column metadata script, and it matches 100% before/after the refactor. The comments in that code make sense to me. 🙏 🚢 🚀
